### ISSUE TITLE
Generar XML

### ISSUE THREAD

**Issue by mauriciobaeza** [UniversoLibreMexicoAC/cfdi-test#4]
Bajo este issue, discutiremos la generación del XML...

**Comment by mauriciobaeza** [comment]
Explico brevemente el estado actual.

- La clase que manejo en Factura Libre es funcional pero esta muy ligada a otros procesos dentro de Factura Libre, mi primer tarea es limpiarla para que sea una clase independiente y ponerla a su consideración.
- La librería usada es ElementTree para todo lo que tiene que ver con XML, la principal razón es que esta dentro del core de Python y no requerimos dependencias externas y es aceptablemente rápida para los XML que se generan.
- Factura Libre soporta varios de los principales complementos de facturación, como: escuelas, donativos, etc... aquí una [lista de los mismos en el SAT](http://www.sat.gob.mx/informacion_fiscal/factura_electronica/Paginas/complementos_factura_cfdi.aspx). Nuestra librería debería, al final, soportarlos todos.

Si vamos documentando todo el proceso y el porque? de cada decision, creo que podríamos llegar a tener una excelente guía para facturación electrónica para cualquier lenguaje.

Un abrazo

**Comment by mauriciobaeza** [comment]
Listo, ahora podemos generar un XML para el caso general, es decir, el más común.

    from cfdi import CFDI
    
    cfdi = CFDI()
    xml = cfdi.get_cfdi(data)

**data** es un diccionario con todos los datos mínimos necesarios para un XML válido técnicamente. He creado, sellado y timbrado un documento con esta clase, solo para validar que se genera correctamente.

Pongo a discución los siguientes puntos:

1. ¿Que tan abstracta debería ser nuestra clase?
1. ¿Solo recibimos datos o también validaremos?

Propongo una reunión de trabajo por mumble.

Un abrazo

**Comment by agb80** [comment]
Buenas, antes que nada feliz inicio de año 2016 y espero que Dios los colme de bendiciones durante este año.

No entiendo bien a que te refieres con "que tan abstracta", en cuanto a la validación, se me ocurre que validar el XML generado contra el xst proporcionado por el SAT sería suficiente, ¿como ves?

**Comment by mauriciobaeza** [comment]
Feliz año para ti también

Me refiero a lo siguiente, por ejemplo, algunas librerías hacen esto:

cfid.EmisorRFC = "blablabla"
cfdi.EmisorNombre = "blablabla"

etc...

otras separan los datos en rubros, por ahora nosotros lo tenemos todo como un gran diccionario, que me parece mucho más flexible, sobre todo por que ya hice pruebas de integrar en un componente que se puede llamar desde otros lenguajes, no todos soportan diccionarios pero todos soportan matrices, así que solo hay que hacer una sola verificación del tipo de datos recibido, en cuanto generé el ejemplo me explicaré mejor...

Por supuesto todo esto que menciono entra en el amplio criterio de desarrollo que cada uno tenemos...

Si, me parece que lo mejor es solo validar el esquema...

Saludos

**Comment by agb80** [comment]
Efectivamente, he visto librerías que separan todos los datos así como dices pero también lo considero poco flexible y extendible. 

Por ejemplo en la nueva especificación del CFDI, ya no es un solo esquema para todos los documentos fiscales sino que son varios, así que tendríamos que crear una subclase para cada documento y a cada una agregar los datos que están pidiendo. 

+1 por el diccionario

**Comment by mauriciobaeza** [comment]
De acuerdo con esto...

Entonces propongo, ya que tenemos un XML basico, terminar el resto de los procesos acordados: sellado, timbrado... para entonces solo regresar al principio a cubrir los diferentes esquemas, pero ya tendríamos el proceso completo para ir validando que son correctos...

¿Te parece?

**Comment by agb80** [comment]
De acuerdo
